Improve API Error Handling

PSalant726 opened the pull request:
Closes #212 
Closes #233 

### Code Changes

* [x] Improve the consistency of DAL function signatures
* [x] Handle exceptions throughout `fidesapi/` using `HTTPException`

### Steps to Confirm

* [ ] Hit all the API endpoints, with reckless abandon

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* [x] Documentation Updated

### Description Of Changes

The [FastAPI documentation on error handling](https://fastapi.tiangolo.com/tutorial/handling-errors) defines the `HTTPException` as the best way to handle errors and return useful information to the caller.

ThomasLaPiana commented:
i'm digging in to this 👍

ThomasLaPiana commented on fidesctl/src/fidesapi/crud.py:
```diff
@@ -160,17 +183,24 @@ async def update(
+
         return query_result
 
-    @router.delete("/{fides_key}", status_code=status.HTTP_204_NO_CONTENT)
```
what's your thinking in changing this to a 200 with a response? I've seen anecdotally that its common to use 204 for deletes and not return anything

PSalant726 replied:
There's definitely some room for opinion with any HTTP status code decision. In this case, I think the two philosophies are:
- Return a 200, because the request was valid and completed successfully. Send a response payload to provide feedback to the caller, ideally including the resource that was deleted.
- Return a 204, because the request was valid and completed successfully, thus removing the resource, which obviously couldn't be fetched afterwards to include in a response.

IMO, the first option is the right choice for this API. The request and database query burdens are low currently, and I would expect that to be true even in a production environment. Because we're not attempting to scale this API to handle millions+ of req/s, we can afford to include a response payload that provides more context and feedback. If that changes later, we can easily switch to a 204/zero-payload response schema (although, of course, it would constitute a breaking change).

ThomasLaPiana replied:
makes sense to me!

ThomasLaPiana commented on fidesctl/src/fidesapi/crud.py:
```diff
@@ -15,13 +15,28 @@
+            "fides_key": fides_key,
+        }
+
+        super().__init__(status.HTTP_422_UNPROCESSABLE_ENTITY, detail=detail)
```
I think a 422 is supposed to be for malformed data specifically, and it is the default error that FastAPI returns when a request comes in and is invalid.

For a true `NotFound`, I think 404 is a better code

PSalant726 replied:
I originally had a 404 response here, but I was convinced to switch to a 422 after re-reading [the spec](https://datatracker.ietf.org/doc/html/rfc4918#section-11.2). It implies to me that because the requester hit a valid route (but included a bad request payload), a 404 would be inappropriate.

ThomasLaPiana replied:
the same spec mentions this though: https://datatracker.ietf.org/doc/html/rfc4918#section-9.6, i.e. after a delete a get of the same resource should throw a 404

It isn't that the request is malformed or had bad data, the resource itself just doesn't exist in the database. 404 can be used for both "this route doesn't exist" and also "this resource doesn't exist"

PSalant726 replied:
Ok, yea I don't feel strongly about returning a 422 for this. I changed it to 404.

ThomasLaPiana reviewed (commented):
Some questions here around status codes, also ran into some uncaught exceptions

ThomasLaPiana commented:
@PSalant726 i went to the docs site and abused the "try it out", and was able to get the unhelpful `Internal Server Error` in these 2 cases:

* Submitted a data category that was invalid (fides_key matched parent_key)
* Submitted a data category that was valid but already existed

If someone is using the CLI these errors would be impossible because the `parse` and `apply` commands would catch them before they hit the server, but since we're here we might as well add those checks and treat this as a zero trust environment

Pydantic should be catching the invalid object error, but for some reason isn't surfacing it. I vaguely recall that that was working before but I'm not sure what changed that could've affected that

ThomasLaPiana commented:
I'd also like more verbose logging by fastapi/uvicorn. If i attach to the logs of the api container, i can see the startup logs (upserts) but I don't see any logs after that related to incoming requests, errors, etc.

PSalant726 commented:
@ThomasLaPiana 
> - Submitted a data category that was invalid (fides_key matched parent_key)
> - Submitted a data category that was valid but already existed

I'll add error handling for these cases, and try to improve the test coverage as well.

> I'd also like more verbose logging by fastapi/uvicorn. If i attach to the logs of the api container, i can see the startup logs (upserts) but I don't see any logs after that related to incoming requests, errors, etc.

I completely agree, but I think unifying our logging practices and improving the output/verbosity is a larger project that deserves it's own issue/PR. I can create one to track that work.

ThomasLaPiana commented:
> @ThomasLaPiana
> 
> > * Submitted a data category that was invalid (fides_key matched parent_key)
> > * Submitted a data category that was valid but already existed
> 
> I'll add error handling for these cases, and try to improve the test coverage as well.
> 
> > I'd also like more verbose logging by fastapi/uvicorn. If i attach to the logs of the api container, i can see the startup logs (upserts) but I don't see any logs after that related to incoming requests, errors, etc.
> 
> I completely agree, but I think unifying our logging practices and improving the output/verbosity is a larger project that deserves it's own issue/PR. I can create one to track that work.

yeah, I think for the CLI itself leaving everything as print/echo is fine, but let's open an issue to make all output in the API be proper logging.

That being said, can you check if fastapi has some kind of logging switch we turned off? I do recall that we used to get much more robust logging. It might have to do with how we're running uvicorn?

PSalant726 commented on fidesctl/src/fidesapi/crud.py:
```diff
@@ -6,30 +6,61 @@
+    """
+
+    def __init__(self, err: SQLAlchemyError) -> None:
+        detail = {"error": str(err.__dict__["orig"])}
```
I don't like this at all. It looks ugly, and more importantly, it leaks implementation details. I struggled mightily to read into the error instance and create something nicer, but it was very challenging.

ThomasLaPiana replied:
i'll take a stab at this over the weekend and see if I can think of something. It never feels good to ship code you have reservations about so let's see if we can get this looking nicer or just stick with the initial fix you made for now and we can revisit this in #241

PSalant726 reviewed (commented):
@ThomasLaPiana I'm still struggling to handle attempts to create "invalid" `data_category` resources. We might need to pair on this next week.

ThomasLaPiana commented:
@PSalant726 i took a shot at it, different approach than yours so lmk what you think. I'm happy with the error-handling as it is now, we can revisit this later but for now I think its sufficient

PSalant726 commented:
@ThomasLaPiana Nice! I just moved the error definitions to their own file. Also, I added a generic `QueryError` just to make sure we're rolling back transactions on failed queries. Like you said, I think we can revisit the messaging in #241.

ThomasLaPiana reviewed (approved):
Great call moving the errors into their own file, we definitely reached that critical mass. Great iteration on making the API less opaque when handling errors!
